### DISCUSSION TITLE
Best way to list keybindings?

### DISCUSSION THREAD

**Discussion by Spitfire1900** [xonsh/xonsh#5298] [Q&A]
I'm aware of `builtins.__xonsh__.shell.prompter.key_bindings.registries[0].bindings` to print the current keybindings (thanks https://github.com/jnoortheen/xontrib-commands/blob/main/xontrib_commands/cmds/report_keys.py#L18), but I have no idea where to look for VI_MODE keybindings, nor do I know if this is the recommended approach to do a keybinding lookup.

**Comment by anki-code** [comment]
You can start [from prompt-toolkit](https://python-prompt-toolkit.readthedocs.io/en/stable/pages/asking_for_input.html#vi-input-mode) because VI_MODE is the feauture of ptk.

**Reply by anki-code** [reply]
May be it will also help - https://github.com/prompt-toolkit/python-prompt-toolkit/issues/560

**Reply by Spitfire1900** [reply]
Hmm. I need the internal xonsh reference object, which appears to be `builtins.__xonsh__.shell.prompter.key_bindings.bindings`, but finding out which VI bindings exist isn't clear from here; all that's listed is emacs and xonsh bindings

**Reply by Spitfire1900** [reply]
[prompt_toolkit/key_binding/bindings/vi.py](https://github.com/prompt-toolkit/python-prompt-toolkit/blob/465ab02854763fafc0099a2e38a56328c1cb0625/src/prompt_toolkit/key_binding/bindings/vi.py) shows the reference vi bindings. Seeing which ones are exposed in Xonsh is a different matter.

e.g. [_accept_completion()](https://github.com/prompt-toolkit/python-prompt-toolkit/blob/465ab02854763fafc0099a2e38a56328c1cb0625/src/prompt_toolkit/key_binding/bindings/vi.py#L545C6-L551) doesn't seem to be available.
